sdk/helper/storagepacker/storagepacker_v2.go

```diff
@@ -0,0 +1,707 @@
+	for _, bucketKey := range s.getAllBaseBucketKeys() {
+		if _, present := s.bucketsCache.Get(bucketKey); !present {
+			lock := locksutil.LockForKey(s.storageLocks, bucketKey)
+			s.bucketsCache.Insert(bucketKey,
+				&LockedBucket{
+					LockEntry: lock,
+					Bucket: &Bucket{
+						Key:       bucketKey,
+						ItemMap:   make(map[string][]byte),
+						HasShards: false,
+					},
+				})
+		}
+	}
+}
+
+func (s *StoragePackerV2) BucketKeys(ctx context.Context) ([]string, error) {
+	var retErr error
+	s.prewarmCache.Do(func() {
+		diskBuckets, err := logical.CollectKeys(ctx, s.BucketStorageView)
+		if err != nil {
+			retErr = err
+			return
+		}
+		retErr = s.preloadBucketsFromDisk(ctx, diskBuckets)
+	})
+	if retErr != nil {
+		return nil, retErr
+	}
+
+	ret := make([]string, 0, 256)
+	s.bucketsCacheLock.RLock()
+	s.bucketsCache.Walk(func(s string, _ interface{}) bool {
+		ret = append(ret, s)
+		return false
+	})
+	s.bucketsCacheLock.RUnlock()
+
+	return ret, nil
+}
+
+// Parse a bucket from storage, but don't add it to the cache here.
+// (We may want to read a v1 bucket and upgrade it?)
+func (s *StoragePackerV2) DecodeBucket(storageEntry *logical.StorageEntry) (*LockedBucket, error) {
+	uncompressedData, notCompressed, err := compressutil.Decompress(storageEntry.Value)
+	if err != nil {
+		return nil, errwrap.Wrapf("failed to decompress packed storage entry: {{err}}", err)
+	}
+	if notCompressed {
+		uncompressedData = storageEntry.Value
+	}
+
+	var bucket Bucket
+	err = proto.Unmarshal(uncompressedData, &bucket)
+	if err != nil {
+		return nil, errwrap.Wrapf("failed to decode packed storage entry: {{err}}", err)
+	}
+
+	cacheKey := s.GetCacheKey(storageEntry.Key)
+	lock := locksutil.LockForKey(s.storageLocks, cacheKey)
+	lb := &LockedBucket{
+		LockEntry: lock,
+		Bucket:    &bucket,
+	}
+	if lb.Key == "" {
+		lb.Key = storageEntry.Key
+	}
+
+	return lb, nil
+}
+
+// Raw access to one of the storage buckets.
+// The caller is responsible for acquiring the lock and verifying that the bucket
+// has not been sharded in the meantime.
+func (s *StoragePackerV2) GetBucket(ctx context.Context, bucketKey string) (*LockedBucket, error) {
+	cacheKey := s.GetCacheKey(bucketKey)
+
+	s.bucketsCacheLock.RLock()
+	_, bucketRaw, found := s.bucketsCache.LongestPrefix(cacheKey)
+	s.bucketsCacheLock.RUnlock()
+
+	if found {
+		return bucketRaw.(*LockedBucket), nil
+	} else {
+		return nil, fmt.Errorf("key %q not found in cache", cacheKey)
+	}
+
+}
+
+func (s *StoragePackerV2) shardBucket(ctx context.Context, bucket *LockedBucket, cacheKey string) error {
+	numShards := int(math.Pow(2.0, float64(s.BucketShardBits)))
+
+	// Create the shards
+	s.Logger.Info("sharding bucket", "bucket_key", bucket.Key, "num_shards", numShards)
+	defer s.Logger.Info("sharding bucket process exited", "bucket_key", bucket.Key)
+
+	shards := make(map[string]*LockedBucket, numShards)
+	for i := 0; i < numShards; i++ {
+		shardKey := fmt.Sprintf("%x", i)
```

Comment:
That's a clever way to generate a shard key. Totally a curiosity question, not a critique at all, I'm simply curious why the base16 of the number is taken rather than the original base10, versus like a base64 or something. Is it because base16 is more space-efficient while requiring less operations to convert than a base64?

Reply:
That was in the original RFC. There's really no reason you couldn't make a radix tree of this sort work with any branching factor you wanted.  4-bit shards = 1 hex character so that's convenient (it lets us just do string manipulation) but not necessary.